pkg/cloudprovider/aws.go

```diff
@@ -1,32 +1,342 @@
+			PrivateIpAddresses: deleteIPs,
+		}
+		_, err = a.client.UnassignPrivateIpAddresses(&inputV4)
+		if err != nil {
+			return err
+		}
+		return a.waitForCompletion(node, awsapi.StringValueSlice(deleteIPs), true)
+	}
 	return nil
 }
 
 func (a *AWS) GetNodeEgressIPConfiguration(node *corev1.Node) ([]*NodeEgressIPConfiguration, error) {
+	instance, err := a.getInstance(node)
+	if err != nil {
+		return nil, err
+	}
+	instanceV4Capacity, instanceV6Capacity, err := a.getInstanceCapacity(instance)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving the instance capacities, err: %v", err)
+	}
+	networkInterfaces, err := a.getNetworkInterfaces(instance)
+	if err != nil {
+		return nil, err
+	}
+	for _, networkInterface := range networkInterfaces {
+		config := &NodeEgressIPConfiguration{
+			Interface: *networkInterface.NetworkInterfaceId,
+		}
+		v4Subnet, v6Subnet, err := a.getSubnet(networkInterface)
+		if err != nil {
+			return nil, fmt.Errorf("error retrieving the network interface subnets, err: %v", err)
+		}
+		config.IFAddr = ifAddr{}
+		if v4Subnet != nil {
+			config.IFAddr.IPv4 = v4Subnet.String()
+		}
+		if v6Subnet != nil {
+			config.IFAddr.IPv6 = v6Subnet.String()
+		}
+		capV4, capV6 := a.getCapacity(instanceV4Capacity, instanceV6Capacity, networkInterface)
+		config.Capacity = capacity{
+			IPv4: capV4,
+			IPv6: capV6,
+		}
+		return []*NodeEgressIPConfiguration{config}, nil
+	}
 	return nil, nil
 }
+
+// Unfortunately the AWS API (WaitUntilInstanceRunning) only handles equality
+// assertion: so on delete we can't specify and assert that the IP which is
+// being removed is completely removed, we are forced to do the inverse, i.e:
+// assert that all IPs except the IP being removed are there (which should
+// anyways be the case). Hence, use our own poller which verifies on ADD that
+// all IPs that we add are assigned to the node, and on DEL that all IPs being
+// removed have been completely removed from the node.
+func (a *AWS) waitForCompletion(node *corev1.Node, ips []string, deleteOp bool) error {
+	return wait.PollImmediate(time.Second*2, time.Minute, func() (done bool, err error) {
+		instance, err := a.getInstance(node)
+		if err != nil {
+			return false, err
+		}
+		sampleIP := ips[0]
+		assignedIPs := []string{}
+		if utilnet.IsIPv6String(sampleIP) {
+			for _, assignedIPv6 := range instance.NetworkInterfaces[0].Ipv6Addresses {
+				if assignedIP := net.ParseIP(*assignedIPv6.Ipv6Address); assignedIP != nil {
+					assignedIPs = append(assignedIPs, assignedIP.String())
+				}
+			}
+		} else {
+			for _, assignedIPv4 := range instance.NetworkInterfaces[0].PrivateIpAddresses {
+				if assignedIP := net.ParseIP(*assignedIPv4.PrivateIpAddress); assignedIP != nil {
+					assignedIPs = append(assignedIPs, assignedIP.String())
+				}
+			}
+		}
+		if !deleteOp {
+			return sets.NewString(assignedIPs...).HasAll(ips...), nil
+		}
+		return !sets.NewString(assignedIPs...).HasAny(ips...), nil
```

Comment:
```
if deleteOp {
        return !sets.NewString(assignedIPs...).HasAny(ips...), nil
} else {
        return sets.NewString(assignedIPs...).HasAll(ips...), nil
}
```

the way it is now suggests that `!deleteOp` is a special case to be handled first, and then `return !sets.NewString(assignedIPs...).HasAny(ips...), nil` is the default behavior. But neither is default or special-case. They're equal.